Update Chapter 6 (#321)
r-causal/causal-inference-in-R: chapters/06-stats-models-ci.qmd

  summarise(estimate = mean(daily - weekly))
```

As opposed to binary and categorical confounders, grouping by bins for continuous confounders does not completely account for the variable; the coarser the bins, the more residual confounding there will be, and the finer the bins, the closer we'll get to the continuous version (but the fewer values we'll have per bin, see @tip-bins).

::: {#tip-bins .callout-tip}

## What would happen if we change the number of bins?

Let's see what happens if we increase the number of bins. In the figure below, we've changed the number of bins from 5 in the example in the text to range from 3 to 20. Notice as we increase the number of bins the bias decreases. 

```{r}
#| code-fold: true
update_bins <- function(bins) {
  satisfaction3 |>
    mutate(num_users_q = ntile(num_users, bins)) |>
    group_by(num_users_q, update_frequency) |>
    summarise(
      avg_satisfaction = mean(satisfaction),
      .groups = "drop"
    ) |>
    ungroup() |>
    pivot_wider(
      names_from = update_frequency,
      values_from = avg_satisfaction
    ) |>
    summarise(
      bins = bins,
      estimate = mean(daily - weekly)
    )
}

> map_dfr(3:20, update_bins) |>

Reviewer: ```suggestion
map(3:20, update_bins) |>
  bind_rows() |>
```

This (or `purrr::list_rbind()`) is the recommended way to do this these days

  ggplot(aes(x = bins, y = abs(estimate))) +
  geom_point() +
  geom_line() +
  labs(y = "Bias", x = "Number of bins")
```

For example, looking at the output below we see that the estimate is much closer to the truth (0) when we have 20 bins compared to 5. 

```{r}
satisfaction3 |>
  mutate(num_users_q = ntile(num_users, 20)) |>
  group_by(num_users_q, update_frequency) |>
  summarise(
    avg_satisfaction = mean(satisfaction),
    .groups = "drop"
  ) |>
  ungroup() |>
  pivot_wider(
    names_from = update_frequency,
    values_from = avg_satisfaction
  ) |>
  summarise(estimate = mean(daily - weekly))
```

As with many good things, however, there is a limit to the utility of increasing the number of bins. For example, let's see what happens if we try to have 30 bins. 

```{r}
satisfaction3 |>
  mutate(num_users_q = ntile(num_users, 30)) |>
  group_by(num_users_q, update_frequency) |>
  summarise(
    avg_satisfaction = mean(satisfaction),
    .groups = "drop"
  ) |>
  ungroup() |>
  pivot_wider(
    names_from = update_frequency,
    values_from = avg_satisfaction
  ) |>
  summarise(estimate = mean(daily - weekly))
```

> The estimate is `NA` because some of our bins didn't have anyone in one of the exposure groups, making their difference inestimable. Now this analysis violates our *positivity* assumption. Here, this is a stochastic violation, it has to do with our sample size, `r scales::comma(n)`, and number of bins, in this case 30. By chance, we ended up with at least one of the 30 bins without anyone in one of the exposure groups, making our causal effect inestimable. This non-parametric method, while flexible, has limitations due to the sample size. Parametric models are useful because they allow us to extrapolate under certain assumptions, which makes them more efficient (assuming the assumptions are true, let's learn more in @sec-parametric). 

Reviewer: ```suggestion
The estimate is `NA` because some of our bins didn't have anyone in one of the exposure groups, making their difference inestimable. Now, this analysis violates our *positivity* assumption. This is a stochastic violation; it has to do with our sample size, `r scales::comma(n)`, and the number of bins, 30. By chance, we ended up with at least one of the 30 bins without anyone in one of the exposure groups, making our causal effect inestimable. This non-parametric method, while flexible, has limitations due to the sample size. Parametric models are useful because they allow us to extrapolate under certain assumptions, which makes them more efficient (assuming the assumptions are true, let's learn more in @sec-parametric). 
```

Author: (fixed this in coming commit)

We need the correct functional form for both the exposure and the confounders.
Modeling this well requires an understanding of the nature of the relationship between these variables and the outcome.

::: {#tip-functional-form .callout-warning}

## Functional form in parametric models

In the text we simulated the relationship between the outcome and the confounders to be linear, that is, it exactly met the assumptions underlying `lm()`, so we got the right answer when we fit our parameteric model. What would happen if our simulation did not match the assumptions underlying `lm()`? Let's take a look.

```{r}
set.seed(1)
satisfaction4 <- tibble(
  # Number of users
  num_users = runif(n, 1, 500),
  # Larger customers more likely to have daily updates
  update_frequency = rbinom(n, 1, plogis(num_users / 100)),
  # non-linear relationship between satisfaction and number of users
  satisfaction = 70 - 0.001 * (num_users-300)^2 - 0.001 * (num_users - 300)^3
) |>
  mutate(
    satisfaction = as.numeric(scale(satisfaction)),
    update_frequency = factor(
      update_frequency,
      labels = c("weekly", "daily")
    )
  )
ggplot(satisfaction4, aes(x = num_users, y = satisfaction)) +

>   geom_point()

Reviewer: This ends up looking like a very thick line. Maybe we should just do `geom_line()` at this point?

```

In the figure above we see that now there is a non-linear relationship between our confounder, number of users, and our outcome, satisfaction. Let's see what happens if we fit an (incorrect) parameteric model to these data.

```{r}
lm(
  satisfaction ~ update_frequency + num_users,
  data = satisfaction4
) |>
  tidy(conf.int = TRUE) |>
  filter(term == "update_frequencydaily") |>
  select(estimate, starts_with("conf"))

> ```

Reviewer: ````suggestion
```

````

> Now our estimates are far off from the truth (which should be zero); the truth is not even contained in the confidence interval. What went wrong? Our parametric model assumed that the functional form of the relationship between the number of users and satisfaction was linear, but we generated it non-linearly. There is a solution that still allows for the use of a parameteric model - if we knew the true functional form, we could use that. Let's see how that looks.

Reviewer: ```suggestion
Our estimates are far from the truth (which should be zero); the truth is not even contained in the confidence interval. What went wrong? Our parametric model assumed that the functional form of the relationship between the number of users and satisfaction was linear, but we generated it non-linearly. There is a solution that still allows for the use of a parametric model; if we knew the true functional form, we could use that. Let's see how that looks.
```

```{r}
lm(
  satisfaction ~ update_frequency + poly(num_users, 3),
  data = satisfaction4
) |>
  tidy(conf.int = TRUE) |>
  filter(term == "update_frequencydaily") |>
  select(estimate, starts_with("conf"))
```

> Beautiful! Now, this model was fit *exactly* as the data were generated, and again we end up with the exact right answer. Often in the real world, we do not know what the data generating mechanism is, but we can still fit flexible parametric models. A great way to do this is through natrual cubic splines.

Reviewer: ```suggestion
Beautiful! Now, this model was fit *exactly* as the data were generated, and again, we ended up with the exact right answer. In the real world, we often do not know the data-generating mechanism, but we can still fit flexible parametric models. A great way to do this is through natural cubic splines.
```

```{r}
lm(
  satisfaction ~ update_frequency + splines::ns(num_users, 3),
  data = satisfaction4
) |>
  tidy(conf.int = TRUE) |>
  filter(term == "update_frequencydaily") |>
  select(estimate, starts_with("conf"))

> ```

Reviewer: ````suggestion
```

````

> We can also use our original non-parametric method, if we stratify this into 20 bins, we also get better estimate.

Reviewer: ```suggestion
We can also use our original non-parametric method. If we stratify this into 20 bins, we also get better estimate.
```

> We can also use our original non-parametric method, if we stratify this into 20 bins, we also get a less biased estimate (i.e. it is very close to the true value, 0).

Reviewer: ```suggestion
We can also use our original non-parametric method; if we stratify this into 20 bins, we also get a less biased estimate (i.e., it is very close to the true value, 0).
```

-   If 50% of the customers are premium and 50% are free, the average effect of switching to daily updates would be:

>     $(0.5 * 5) + (0.5 * -5) = 0$

Reviewer: these deletions also seem unintentional

Author: indeed 🙈 it seems RStudio did this to me??

Here's a brief summary of some of the unconfoundedness methods we'll cover and what they do.

-   *Unconfoundedness methods*

>     -   **Inverse probability weighting** (propensity score weighting): Using a propensity score (predicted probability of treatment), we reweight units to create a pseudo-population where exchangeability holds. Extends to time-varying treatments.

Reviewer: Are these deletions intentional?

Reviewer: Looking at the rendered versions, I'm going to guess no because they look weird now

Author: no again! will fix.
